hare/broker.go

```diff
@@ -104,49 +112,72 @@ func (b *Broker) eventLoop() {
 				b.Warning("Message validation failed: InnerMsg is nil")
 				continue
 			}
-			expInstId := b.maxReg + 1 //  max expect current max + 1
+
 			msgInstId := InstanceId(hareMsg.InnerMsg.InstanceId)
-			// far future unregistered instance
-			if msgInstId > expInstId {
-				b.Warning("Message validation failed: InstanceId. Max: %v Actual: %v", expInstId, msgInstId)
+			state, exist := b.layerState[msgInstId]
+			if exist && state == invalid { // invalid instance, ignore
+				b.Info("%v instance is invalid. Ignoring message", msgInstId)
 				continue
 			}
 
-			// near future
-			if msgInstId == expInstId {
-				futureMsg = true
+			c, ok := b.outbox[msgInstId]
+			if !ok { // unknown instance, maybe just an early msg
+				if b.latestLayer+1 != msgInstId { // not an early msg
+					// ignore msg
+					b.Warning("Message validation failed: InstanceId. Expected: %v Actual: %v", b.latestLayer+1, msgInstId)
+					continue
+				}
+
+				// check synchronicity
+				if !b.isSynced(msgInstId) {
+					// not synced, we will ignore messages for this instance from now on
+					b.Info("layer %v is not synced, ignoring message", msgInstId)
+					continue
+				}
+
+				// early msg, should validate & buffer
+				// but we do it later to avoid duplication
 			}
 
+			// the msg is either early or has instance
+
+			// create msg
 			iMsg, err := newMsg(hareMsg, b.stateQuerier, b.layersPerEpoch)
 			if err != nil {
 				b.Warning("Message validation failed: could not construct msg err=%v", err)
 				continue
 			}
 
+			// validate msg
 			if !b.eValidator.Validate(iMsg) {
 				b.Warning("Message validation failed: eValidator returned false %v", hareMsg)
 				continue
 			}
 
-			// validation passed
+			// validation passed, report
 			msg.ReportValidation(protoName)
 
-			c, exist := b.outbox[msgInstId]
-			if exist {
-				// todo: err if chan is full (len)
-				c <- iMsg
-				continue
-			} else { // InnerMsg arrived before registration
-				futureMsg = true
-			}
-
-			if futureMsg {
-				if _, exist := b.pending[msgInstId]; !exist {
+			// early msg
+			if !ok {
+				if _, exist := b.pending[msgInstId]; !exist { // create buffer if first msg
 					b.pending[msgInstId] = make([]*Msg, 0)
 				}
+				if len(b.pending[msgInstId]) == InboxCapacity { // buffer is full
+					b.Error("Reached %v pending messages. Ignoring message for layer %v sent from %v",
+						InboxCapacity, msgInstId, iMsg.PubKey.ShortString())
+					continue
+				}
 				b.pending[msgInstId] = append(b.pending[msgInstId], iMsg)
+				continue
+			}
+
+			// has instance, just send
+			if !ok {
```

Comment:
don't understand this code... It seems that it is unreachable (if ok is false you'll hit the continue in the previous if)

Reply:
Done